gadget/mountedfilesystem.go

```diff
@@ -0,0 +1,184 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type MountedFilesystemWriter struct {
+	contentDir string
+	ps         *PositionedStructure
+}
+
+// NewMountedFilesystemWriter returns a writer capable of deploying provided
+// structure, with content of the structure stored in the given root directory.
+func NewMountedFilesystemWriter(contentDir string, ps *PositionedStructure) (*MountedFilesystemWriter, error) {
+	if ps == nil {
+		return nil, fmt.Errorf("internal error: *PositionedStructure is nil")
+	}
+	if ps.IsBare() {
+		return nil, fmt.Errorf("structure %v has no filesystem", ps)
+	}
+	if contentDir == "" {
+		return nil, fmt.Errorf("internal error: gadget content directory cannot be unset")
+	}
+	fw := &MountedFilesystemWriter{
+		contentDir: contentDir,
+		ps:         ps,
+	}
+	return fw, nil
+}
+
+func prefixPreserve(dstDir string, preserve []string) []string {
+	preserveInDst := make([]string, len(preserve))
+	for i, p := range preserve {
+		preserveInDst[i] = filepath.Join(dstDir, p)
+	}
+	sort.Strings(preserveInDst)
+
+	return preserveInDst
+}
+
+// Write writes structure data into provided directory. All existing files are
+// overwritten, unless their paths, relative to target directory, are listed in
+// the preserve list.
+func (m *MountedFilesystemWriter) Write(whereDir string, preserve []string) error {
+	if whereDir == "" {
+		return fmt.Errorf("internal error: destination directory cannot be unset")
```

Comment:
Should we have a wrapper for making internal errors? It seems we have quite a few of those and they are just a typo away from `intrenal error`

Reply:
wouldn't how spell check test find that?  anyway -0 on a wrapper for this